### ISSUE TITLE
how to represent extension target attributes

### ISSUE THREAD

**Issue by dret** [dret/I-D#74]
as mentioned in https://github.com/dret/I-D/issues/71#issuecomment-321512447, there should be a section about how to represent extension target attributes, i.e. all those target attributes that are not defined by RFC 5988bis, but may be found in links and may have to be mapped to JSON. it probably would make sense to have an individual section about this.

**Comment by dret** [comment]
it probably would be useful to interlink this section with the sections describing the representation of the existing target attributes. this way, people can start either way (reading about how existing target attributes are represented, or what the patterns are for extension target attributes) and find their way around.

**Comment by dret** [comment]
one of the challenges here is that when looking at a specific linkset that uses extension target attributes, it is impossible to tell how these are defined. we essentially have two different audiences here:
* when taking linkset instances, how to handle extension target attributes.
* when defining target attributes, maybe take into account how they get represented in JSON (where simply repeating a target attribute is not a syntax option that is available).
looking at it from this side of things, it may even be worth it to separate these two cases.

**Comment by dret** [comment]
thinking more about this, it seems to me that we cannot get around addressing things as proposed in the last comment (instance and schema based). let's assume we have a linkset like this:

```
<http://example.com/link> ; rel="linkset" ; extension="value"
```

how is this going to be represented in JSON? probably simply `"extension" : "value"`, but we don't really know whether `extension` is just a single value, follows the pattern of #71 (i.e., may repeat on other instances), or follows the pattern of #73 (i.e., may contain repeated values based on some microsyntax).
absent any info about the *target attribute definition*, what do we say has to be done? the last update seems to imply that we MUST ignore `extension`, but that seems unnecessarily draconian. wouldn't it make more sense to define `"extension" : "value"` and say that repeating values MUST be ignored, unless there is additional knowledge about the target attribute definition?

**Comment by dret** [comment]
additional thought of how to approach this:
iff the JSON serializer has schema-based knowledge on how specific extension target attributes have to be handled, then it can apply that knowledge. knowing the predefined target attributes of RFC 5988bis then simply is a special case of this: because of this, the serializer knows to ignore duplicate `hreflang` (https://github.com/dret/I-D/commit/4afd0df8ca5d5f35ff468b8f8a1881a1bd3e0ac1). because of this, a serializer might decide to serialize `rel='x y'` (#73) as two JSON link objects, or not (as suggested in https://github.com/dret/I-D/commit/83a46479a925ceea93048d5adbd7bb94f3ebacc3)

**Comment by dret** [comment]
not sure if this belong here or is a separate issue: if we look at "schema-based knowledge", i.e. a JSON serializer knowing a particular extension target attribute: would we ever allow target attribute values to be anything other than a string? my feeling is that we shouldn't allow this, making sure that whatever structure extension target attributes might have has to be represented in a microsyntax that can be stuffed into a string. at some level, this is what #73 is all about, where the desire to represent multiple relation types might lead us to say we don't support it, we support it via `"rel": "x y"` but would probably not lead us to defining `rel` as becoming `"rel": [ "x", "y" ]`.

**Comment by hvdsomp** [comment]
While I am sympathetic to some considerations above, I think the JSON serialization is seriously constrained when:
- Not allowing to express multiple values for hreflang, which is defined by the serialization of RFC5899bis and hence should be known by all clients
- Not allowing any repeatable attributes
- Allowing repeatable attributes but mandating a space-delimited microsyntax approach for them, which means having to interpret a space in any extension attribute value as separating multiple values

I think a core consideration mentioned in the above, which actually is expressed in the current version of the I-D, is that implementations must ignore attributes they do not understand. Not understanding an attribute includes not understanding its semantics, cardinality and content model. Understanding an attribute means understanding its semantics, cardinality and content model. That is the case for both implementations that write link sets and for those that consume them.

I think that [Siren](https://github.com/kevinswiber/siren), which you pointed out to me, has an attractive approach to deal with repeatability: turn it into an array. 

So, let's look at our challenging cases:

* How to handle a link object that expresses multiple links with same link context and link target
`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  ["start" , "index"], 
    "type" : "text/html"} ]`

* How to handle repeatable attributes of the token variant (e.g. hreflang) 
`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  ["start"], 
    "type" : "text/html",
    "hreflang": ["en" , "de"] } ]`

* How to handle non-repeatable attributes of the token* variant (e.g. title*) 
`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  ["start"], 
    "type" : "text/html",
    "title": "Good morning",
    "title*": "'de'Güten Morgen" } ]`

* How to handle repeatable attributes of the token* variant (e.g. extension-attribute*) 
`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  ["start"], 
    "type" : "text/html",
    "extension-attribute*": ["'nl'dit en dat" , "'fr'ceci et cela" ] } ]`

To summarize, three serialization rules are underlying the above examples:
- Ignore attributes you don't understand 
- For repeatable attributes use an array to express values
- For token* attributes, represent an individual value as: 'language-tag'content

**Comment by asbjornu** [comment]
I agree with @hvdsomp that using JSON arrays instead of microsyntax is preferable. In the same vein I'm not a fan of the language tag microsyntax. How about turning the value of language tagged attributes into an object where each key is the language tag as such?

```json
[{
    "href": "http://example.com/foo",
    "anchor": "http://example.net/bar",
    "rel": ["start"],
    "type": "text/html",
    "extension-attribute*": {
        "nl": "dit en dat",
        "fr": "ceci et cela"
    }
}]
```

It's easy enough to have a different content model for `extension-attribute` and `extension-attribute*`, so I don't see why we need a microsyntax for the `*` attribute to somehow have the seemingly (but in reality not) same content model for the non-language tagged attribute.

**Comment by dret** [comment]
On 2017-08-15 08:09, Asbjørn Ulsberg wrote:
> It's easy enough to have a different content model for
> |extension-attribute| and |extension-attribute*|, so I don't see why we
> need a microsyntax for the |*| attribute to somehow have the seemingly
> (but in reality not) same content model for the non-language tagged
> attribute.

i agree that it's possible to move to a JSON model that is more 
"diverse" than a simple key/string-value model. however, there are 
certain implications for how that limits intermediaries to process 
linksets when they do not have built-in knowledge of all link parameters.

i have created https://github.com/dret/I-D/issues/76 and 
https://github.com/dret/I-D/issues/77 to focus on these issues. 
resolution of https://github.com/dret/I-D/issues/73 and 
https://github.com/dret/I-D/issues/74 probably will be based on the 
discussions for these more general design issues.

**Comment by dret** [comment]
> While I am sympathetic to some considerations above, I think the JSON
> serialization is seriously constrained when:
>
>   * Not allowing to express multiple values for hreflang, which is
>     defined by the serialization of RFC5899bis and hence should be known
>     by all clients

https://github.com/dret/I-D/issues/71

>   * Not allowing any repeatable attributes

https://github.com/dret/I-D/issues/76#issuecomment-322577333

>   * Allowing repeatable attributes but mandating a space-delimited
>     microsyntax approach for them, which means having to interpret a
>     space in any extension attribute value as separating multiple values
> I think a core consideration mentioned in the above, which actually is
> expressed in the current version of the I-D, is that implementations
> must ignore attributes they do not understand.

it's true that with such a requirement things would become much easier. 
however, it seems that such a requirement does go against the grain of 
intermediaries not dropping things (much less requiring them to drop 
them), which is discussed in https://github.com/dret/I-D/issues/76

**Comment by hvdsomp** [comment]
I think that, in essence, we are looking for a JSON serialization that allows transforming from the native serialization to JSON and from JSON to native without loss of expressiveness. Discussions show this is challenging, among others due to repeatability of serialization-defined attributes, and unknown repeatability of extension attributes. I took a first stab, the other day, and came up with a proposal that allowed:
- blindly transforming from JSON to native without any loss 
- transforming from native to JSON only for those attributes that an application “understood”

The latter leads to the problem of clients having to ignore attributes they don’t understand. Below, I am having another try, much inspired by the suggestion @mamund made in #76, and I think I have now arrived at an approach that allows:
- blindly transforming from JSON to native without any loss 
- blindly transforming from native to JSON without any loss

The proposed approach handles serialization-defined and extension attributes differently:
- It leverages the fact that serialization-defined attributes should be known to clients and hence gives them a special treatment. It makes them look like first class citizens.
- It considers the fact that extension attributes may be unknown to the client and treats them uniformly. As second class citizens ;-)

But, as @mamund indicated, the uniform treatment could also be applied to serialization-defined attributes. I’m not too fond of doing that, however.

Here goes:

1. Serialization-defined attributes

1.1. How to handle a link object that expresses multiple links with same link context and link target (i.e. how to handle multiple rels)

There are two choices: an array approach or a microsyntax approach. We need to pick one. If we choose the array approach, it must always be used, even if there is only a single rel.

1.1.1. array approach

`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  ["start" , "index"] } ]`

1.1.2. microsyntax approach

`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  "start index" } ]`

1.2. How to handle non-repeatable attributes of the token variant defined by RFC5899bis (e.g. type, title)

For brevity, only one example is shown, i.e. using the microsyntax approach of 1.1.2. for rel. 

`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  "start", 
    "type" : "text/html",
    "title": "Good morning" } ]`

1.3. How to handle repeatable attributes of the token variant defined by RFC5899bis (i.e. hreflang)

There are two choices: an array approach or a microsyntax approach. We need to pick one and the choice should be aligned with the one we make for 1.1.. If we choose the array approach, it must always be used, even if there is only a single hreflang.

1.3.1. array approach
 
`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  ["start"], 
    "type" : "text/html",
    "hreflang": ["en" , "de"] } ]`

1.3.2. microsyntax approach

`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  ["start"], 
    "type" : "text/html",
    "hreflang": "en de" } ]`

1.4. How to handle non-repeatable attributes of the token* variant defined by RFC5899bis (i.e. title*)

For brevity, only one example is shown, i.e. using the microsyntax approach of 1.1.2. for rel.

`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  "start", 
    "type" : "text/html",
    "title": "Good morning",
    "title*": "'de'Güten Morgen" } ]`

1.5. How to handle repeatable attributes of the token* variant defined by RFC5899bis: does not exist

2. Extension attributes

Here, we use the approach suggested by @mamund in #74

2.1. How to handle non-repeatable extension attributes of the token variant

For brevity, only one example is shown, i.e. using the micro syntax approach of 1.1.2. for rel.

`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  "start", 
    "type" : "text/html",
    "extension-attributes”: [ {“name”:”xtension1”,”value":"123"}, {"name":"xtension2”,”value":"abc"} ] } ]`

2.2. How to handle repeatable extension attributes of the token variant

For brevity, only one example is shown, i.e. using the microsyntax approach of 1.1.2. for rel.

`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  "start", 
    "type" : "text/html",
    "extension-attributes”: [ {“name”:”xtension1”,”value":"123"}, {"name":"xtension1”,”value":"abc"} ] } ]`

2.3. How to handle non-repeatable extension attributes of the token* variant

For brevity, only one example is shown, i.e. using the micro syntax approach of 1.1.2. for rel.

`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  "start", 
    "type" : "text/html",
    "extension-attributes”: [ {“name”:”xtension1*”,”value":"'de'Güten Morgen"}, {"name":"xtension2*”,”value”:”’fr’abc"} ] } ]`

2.4. How to handle repeatable extension attributes of the token* variant

For brevity, only one example is shown, i.e. using the micro syntax approach of 1.1.2. for rel.

`[ { "href" : "http://example.com/foo",
    "anchor" : "http://example.net/bar",
    "rel"  :  "start", 
    "type" : "text/html",
    "extension-attributes”: [ {“name”:”xtension1*”,”value":"'de'Güten Morgen"}, {"name":"xtension1*”,”value”:”’fr’abc"} ] } ]`

**Comment by dret** [comment]
i like some of the patterns in there, such as using arrays for repeatable values. but again, this all depends on knowing the schema: if i do not know the attribute, i do not know whether it's repeatable or not, and thus cannot represent it in any way.
after discussing this with @mamund, here's an approach that does not require processors to be schema-aware, and still works: every link parameter that is not predefined is always represented as an array of strings. this way, we don't have to distinguish representations based on schema knowledge, and can safely represent everything (at the expense of a bit more JSON structure than we might need in a schema-optimized model).

**Comment by asbjornu** [comment]
> every link parameter that is not predefined is always represented as an array of strings.

👍  I'd be fine with that.

**Comment by dret** [comment]
On 2017-08-16 14:47, Asbjørn Ulsberg wrote:
>     every link parameter that is not predefined is always represented as
>     an array of strings.
>
> 👍 I'd be fine with that.

good to hear, thanks for the vote!

**Comment by hvdsomp** [comment]
I'm sorry but I'm truly lost now. 

* I really thought we can expect understanding of serialization-defined attributes.  If that assumption is correct, then my proposal works. If not, then we can use the approach shown for extension attributes in all cases.
* For attributes not defined by the serialization, for which we can't expect an understanding, ZERO knowledge is required. Because the proposed serialization is EXACTLY the same in all cases. It may not look that way, because the examples are broken down by theoretical possible use cases. But, if you look closely for a moment, the approach is the same for ALL extension attributes. 

So: There is no schema knowledge required. Or to put it differently:
A. schema knowledge is required in cases where such knowledge can be expected - serialization defined attributes.
B. no schema knowledge is required for extension attributes. zero. zilch. nada. That's the beauty of the approach proposed by @mamund
C. if A is unacceptable (!?) then the B approach can be used for serialization-defined attributes too.

**Comment by dret** [comment]
On 2017-08-16 04:05, Herbert Van de Sompel wrote:
> 1.1. How to handle a link object that expresses multiple links with same
> link context and link target (i.e. how to handle multiple rels)
>
> There are two choices: an array approach or a microsyntax approach. We
> need to pick one. If we choose the array approach, it must always be
> used, even if there is only a single rel.
>
> 1.1.1. array approach
>
> |[ { "href" : "http://example.com/foo", "anchor" :
> "http://example.net/bar", "rel" : ["start" , "index"] } ]|
>
> 1.1.2. microsyntax approach
>
> |[ { "href" : "http://example.com/foo", "anchor" :
> "http://example.net/bar", "rel" : "start index" } ]|

that's what #73 is about.

> 1.3. How to handle repeatable attributes of the token variant defined by
> RFC5899bis (i.e. hreflang)
>
> There are two choices: an array approach or a microsyntax approach. We
> need to pick one and the choice should be aligned with the one we make
> for 1.1.. If we choose the array approach, it must always be used, even
> if there is only a single hreflang.
>
> 1.3.1. array approach
>
> |[ { "href" : "http://example.com/foo", "anchor" :
> "http://example.net/bar", "rel" : ["start"], "type" : "text/html",
> "hreflang": ["en" , "de"] } ]|
>
> 1.3.2. microsyntax approach
>
> |[ { "href" : "http://example.com/foo", "anchor" :
> "http://example.net/bar", "rel" : ["start"], "type" : "text/html",
> "hreflang": "en de" } ]|

that's what #71 is about. my vote is for deciding both #71 and #73 based 
on community feedback, and using a consistent approach.

> 1.4. How to handle non-repeatable attributes of the token* variant
> defined by RFC5899bis (i.e. title*)
>
> For brevity, only one example is shown, i.e. using the microsyntax
> approach of 1.1.2. for rel.
>
> |[ { "href" : "http://example.com/foo", "anchor" :
> "http://example.net/bar", "rel" : "start", "type" : "text/html",
> "title": "Good morning", "title*": "'de'Güten Morgen" } ]|

what about not keeping the RFC 5987 microsyntax, and instead making 
language and content explicit? #67 is about title/title* issues.

>  2. Extension attributes
>
> Here, we use the approach suggested by @mamund
> <https://github.com/mamund> in #74 <https://github.com/dret/I-D/issues/74>
>
> 2.1. How to handle non-repeatable extension attributes of the token variant
>
> For brevity, only one example is shown, i.e. using the micro syntax
> approach of 1.1.2. for rel.
>
> |[ { "href" : "http://example.com/foo", "anchor" :
> "http://example.net/bar", "rel" : "start", "type" : "text/html",
> "extension-attributes”: [ {“name”:”xtension1”,”value":"123"},
> {"name":"xtension2”,”value":"abc"} ] } ]|

what about just using `"xtension1": [ "123" ]`? seems shorter and more 
in line with JSON idioms, and no need to use an extra wrapper?

> 2.2. How to handle repeatable extension attributes of the token variant
>
> For brevity, only one example is shown, i.e. using the microsyntax
> approach of 1.1.2. for rel.
>
> |[ { "href" : "http://example.com/foo", "anchor" :
> "http://example.net/bar", "rel" : "start", "type" : "text/html",
> "extension-attributes”: [ {“name”:”xtension1”,”value":"123"},
> {"name":"xtension1”,”value":"abc"} ] } ]|

this would then simply become `"xtension1": [ "123" , "abc" ]`

> 2.3. How to handle non-repeatable extension attributes of the token* variant
>
> For brevity, only one example is shown, i.e. using the micro syntax
> approach of 1.1.2. for rel.
>
> |[ { "href" : "http://example.com/foo", "anchor" :
> "http://example.net/bar", "rel" : "start", "type" : "text/html",
> "extension-attributes”: [ {“name”:”xtension1*”,”value":"'de'Güten
> Morgen"}, {"name":"xtension2*”,”value”:”’fr’abc"} ] } ]|
>
> 2.4. How to handle repeatable extension attributes of the token* variant
>
> For brevity, only one example is shown, i.e. using the micro syntax
> approach of 1.1.2. for rel.
>
> |[ { "href" : "http://example.com/foo", "anchor" :
> "http://example.net/bar", "rel" : "start", "type" : "text/html",
> "extension-attributes”: [ {“name”:”xtension1*”,”value":"'de'Güten
> Morgen"}, {"name":"xtension1*”,”value”:”’fr’abc"} ] } ]|

same comments as above, both regarding microsyntax for RFC 5987 style 
attributes, and handling repeatability of attributes.

**Comment by ibnesayeed** [comment]
> @dret: after discussing this with @mamund, here's an approach that does not require processors to be schema-aware (#76), and still works: **every link parameter that is not predefined is always represented as an array of strings**. this way, we don't have to distinguish representations based on schema knowledge, and can safely represent everything (at the expense of a bit more JSON structure than we might need in a schema-optimized model).

Are we getting any advantage by splitting a value string into an array of tokens for unknown attributes? Should we not leave the attribute alone as is when we don't understand it to better conform to "implementations must ignore attributes they do not understand"? It also depends on what is the original representation from which the conversion is taking place, because there is limit how much structure can be serialized in the Link format, so out of the band knowledge is necessary to understand the attribute. Although, JSON can hold more structure, but making it the primary format would yield in many lossy conversions. Imagine what would happen if there is an extension attribute that accepts free form text (with spaces) as values, but it is also repeatable.

It is an inherently difficult problem which we might not have a perfect solution for. So far, I have liked the approach summarized by @hvdsomp in his last detailed comment. We can perhaps improve a few things in it. However, in general we might end up relying on describing some best practices and conventions for extension attributes to minimize the possibilities of lossy conversions.

**Comment by ibnesayeed** [comment]
@dret, I might have misunderstood your point when you said, "every link parameter that is not predefined is always represented as an array of strings". Given an unknown attribute `foo-attribute="bar baz"`, if you convert it as `"foo-attribute": ["bar", "baz"]` then my above comment applies. However, if you meant `"foo-attribute": ["bar baz"]` then I am OK with this choice.

**Comment by dret** [comment]
On 2017-08-16 16:13, Sawood Alam wrote:
> @dret <https://github.com/dret>, I might have misunderstood your point
> when you said, "every link parameter that is not predefined is always
> represented as an array of strings". Given an unknown attribute
> |foo-attribute="bar baz"|, if you convert it as |"foo-attribute":
> ["bar", "baz"]| then my above comment applies. However, if you meant
> |"foo-attribute": ["bar baz"]| then I am OK with this choice.

thanks for the feedback and thinking through this with us. it is most 
definitely the second version, i.e. `"foo-attribute": ["bar baz"]`, 
because we shouldn't be parsing values (with the exception of RFC 5987 
style ones, but that is a rather exotic case, see #78).

keep in mind, though, that if the link has `foo-attribute="bar 
baz";foo-attribute="gnu gnat"` on it, then we will end up with 
`"foo-attribute": ["bar baz", "gnu gnat"]` (as suggested in 
https://github.com/dret/I-D/issues/74#issuecomment-322914174).

**Comment by dret** [comment]
suggested resolution for this issue (based on the suggested resolution for #76):

- parameters are treated uniformly as no schema knowledge is required or used.

- each parameter is represented by an array of strings. multiple occurrences of the same parameter on the same link result in multiple string in the parameter value array. (this is consistent with the suggested solutions for #71, #73, and #78)

native: `<http://example.org/>; rel="start"; param="x"`
JSON: `[ { "href":"http://example.org/", "rel":"start", "param":["x"] } ]`

native: `<http://example.org/>; rel="start"; param="x y"`
JSON: `[ { "href":"http://example.org/", "rel":"start", "param":["x y"] } ]`

native: `<http://example.org/>; rel="start"; param="x"; param="y"`
JSON: `[ { "href":"http://example.org/", "rel":"start", "param":["x","y"] } ]`

**Comment by hvdsomp** [comment]
I can definitely agree with this proposal, especially because it yields uniformity regarding handling repeatability for both serialization-defined and extension attributes. It's also shorter and simpler. 

I think it would be nice to make the proposed syntax explicit by means of examples for all the possible cases (1.1. through 2.4.) as listed above.

**Comment by hvdsomp** [comment]
I think I should be a bit more explicit regarding this bit:

`because it yields uniformity regarding handling repeatability for both serialization-defined and extension attributes` 

I meant that if we choose for this approach for extension attributes:

`"param":["x","y"]`

then it seems rather logical to choose for the same approach for the serialization-defined attributes above (cases 1.1. and 1.3.) by using the array variation, not the microsyntax variation:

`"rel" : ["start" , "index"] `
`"hreflang": ["en" , "de"]`

I understand these cases could be handled differently and that handling them as microsyntax might align better with existing practice. Then again, I need to note that the approach proposed in #78 to handle x* attributes also differs from existing practice:

`native: <http://example.org/>; rel="next"; x*="UTF-8'de'n%c3%a4chstes%20Kapitel"
JSON: [ { "href":"http://example.org/", "rel":"next", "x*":[["nächstes Kapitel", "de"]] } ] `

**Comment by dret** [comment]
On 2017-08-17 01:46, Herbert Van de Sompel wrote:
> I think it would be nice to make the proposed syntax explicit by means 
> of examples for all the possible cases (1.1. through 2.4.) as listed above.

we'll figure out a way how to specify/document the syntax through prose 
and examples. but if we agree on the syntax itself, then this issue can 
be closed. sweet!

**Comment by hvdsomp** [comment]
Definitely easier to go check n issues instead of having examples of the evolving ideas regarding serialization in a single spot ;-)

**Comment by dret** [comment]
done.
